source/ui/dataman/AMScanThumbnailGridInputManager.cpp

```diff
@@ -0,0 +1,221 @@
+					// This is the first time this hover that we've updated the
+					// position
+					lastHoverUpdatePositionX_ = positionX;
+				}
+
+				int travelDistanceX = int(std::abs(double(positionX - lastHoverUpdatePositionX_)));
+				if(travelDistanceX > HOVER_TRAVEL_UPDATE_DISTANCE) {
+					// We've moved far enough that we need to emit a hover travel update
+					emit hoverMove(itemIndex, positionX, positionY);
+					lastHoverUpdatePositionX_ = positionX;
+				}
+			}
+		} else {
+			lastMouseOverItemIndex_ = itemIndex;
+		}
+	}
+}
+
+void AMScanThumbnailGridInputManager::mouseReleasedAt(int positionX, int positionY, Qt::MouseButtons mouseButtons, Qt::KeyboardModifiers keys)
+{
+
+	if(startPositionX_ == -1 && startPositionY_ == -1) {
+		// No interaction is currently being performed. Just return.
+
+		return;
+	}
+
+	if(!mouseButtons & Qt::LeftButton) {
+		// Left button clicked
+
+		if(dragInProgress_) {
+			// Drag has stopped here - We don't accept drags, so nothing doing
+
+			resetInteraction();
+		} else {
+			// No Drag is happening
+
+			int travelDistanceX = int(std::abs(double(positionX - startPositionX_)));
+			int travelDistanceY = int(std::abs(double(positionY - startPositionY_)));
+			if(travelDistanceX > TRAVEL_TOLERANCE ||
+					travelDistanceY > TRAVEL_TOLERANCE) {
+				// We've moved far enough that this must be a rectangle, not a point, selection
+
+				if(keys & Qt::ShiftModifier || keys & Qt::ControlModifier) {
+					// Shift or Control is held
+
+					QRect finalSelectionRectangle(startPositionX_, startPositionY_, positionX - startPositionX_, positionY - startPositionY_);
+					emit selectionRectangleEnded(finalSelectionRectangle.normalized(), QItemSelectionModel::Select);
+					resetInteraction();
+				} else {
+					// No key modifier is held
+
+					QRect finalSelectionRectangle(startPositionX_, startPositionY_, positionX - startPositionX_, positionY - startPositionY_);
+					emit selectionRectangleEnded(finalSelectionRectangle.normalized(), QItemSelectionModel::ClearAndSelect);
+					resetInteraction();
+				}
+			} else {
+				// Just a single point click
+
+				if(keys & Qt::ControlModifier) {
+					// Control held - toggle the specified index without clearing
+
+					emit itemSelected(lastClickedItemIndex_, QItemSelectionModel::Toggle);
+				} else if(keys & Qt::ShiftModifier) {
+					// Shift held - extend the current selection to the last clicked index
+
+					emit selectionExtended(lastClickedItemIndex_);
+				} else {
+					// Nothing held - clear current selection and select the lastClickedItemIndex_
+
+					emit itemSelected(lastClickedItemIndex_, QItemSelectionModel::ClearAndSelect);
+				}
+				resetInteraction();
+				doubleClickTimer_.start(DOUBLE_CLICK_INTERVAL_MS);
+			}
+		}
+	} else if(!mouseButtons & Qt::RightButton){
+		// Right button clicked
+
+		if(dragInProgress_) {
+			// Drag is in progress need to cancel it
+		} else {
+			int travelDistanceX = int(std::abs(double(positionX - startPositionX_)));
```

Comment:
qAbs is much better, as its a templated function which means I don't have to the that ghastly cast to double then cast to int.